manifold_airflow_dags/manifold_db_sync_dag.py

```diff
@@ -0,0 +1,173 @@
+MANIFOLD_DUMP_DATABASE_INTERVAL = \
+        Variable.get("MANIFOLD_DUMP_DATABASE_SCHEDULE_INTERVAL", "@never")
+AIRFLOW_S3_CONN_ID = "AIRFLOW_S3"
+AIRFLOW_S3_PREFIX = "manifold/"
+AIRFLOW_DATA_BUCKET = Variable.get("AIRFLOW_DATA_BUCKET")
+
+DEFAULT_ARGS = {
+    'owner': 'airflow',
+    'start_date': datetime(2019, 5, 28),
+    'email': ['steven.ng@temple.edu'],
+    'email_on_failure': False,
+    'email_on_retry': False,
+    'retries': 0,
+    'on_failure_callback': slackpostonfail,
+    'retry_delay': timedelta(minutes=5),
+}
+
+DAG = airflow.DAG(
+    'manifold_database_sync',
+    default_args=DEFAULT_ARGS,
+    catchup=False,
+    max_active_runs=1,
+    schedule_interval=MANIFOLD_DUMP_DATABASE_INTERVAL,
+)
+
+#
+# CREATE TASKS
+#
+# Tasks with all logic contained in a single operator can be declared here.
+# Tasks with custom logic are relegated to individual Python files.
+
+# 1) Find the latest db dump in s3
+# 2) Copy the Dump from s3 to the downstream DB Server
+# 3) Drop and pg_restore the DB from the dump on the downstream server
+# 4) Run a DB migration from the downstream App server to bring the DB back into shape
+# 5) restart the rails app on the downstream server
+# 6) Do some cleanup of the files on the downstream DB server
+#
+
+
+def newest_key_with_prefix(aws_conn_id, bucket, prefix):
+    client =  S3Hook(aws_conn_id=aws_conn_id).get_conn() # get the underlying boto3 client object
+    s3_keys = client.list_objects_v2(Bucket=bucket, Prefix=prefix).get("Contents", [])
+    in_date_order = sorted(s3_keys, key= lambda k: k['LastModified'])
+    path = in_date_order[-1]['Key']
+    return {"full_path": path, "filename": path.split('/')[-1]}
+
+DETERMINE_S3_DB_DUMP_FILE = PythonOperator(
+    task_id="get_latest_db_dump_in_s3",
+    python_callable=newest_key_with_prefix,
+    provide_context=False,
+    op_kwargs={
+        "aws_conn_id": AIRFLOW_S3_CONN_ID,
+        "bucket": AIRFLOW_DATA_BUCKET,
+        "prefix": AIRFLOW_S3_PREFIX,
+    },
+    dag=DAG)
+
+POST_SLACK = PythonOperator(
+    task_id="slack_post_succ",
+    python_callable=slackpostonsuccess,
+    provide_context=True,
+    dag=DAG)
+
+# We use the restore user for this because the restore user does not exist in the prod DB
+# This prevents us from accidentally running the restore on prod db.
+
+
+# We need to drop all connections to the manifold DB before we can drop the DB. But we don't want to drop the connection from the 
+# psql command doing the dropping or else the command fails.
+# So we
+# 1. drop connections to the db
+# 2. drop the db
+# 3. run pg_restore in a way that recreates the db and tables.
+# TODO - pg_restore returns some ignorable warnings that cause the command rc to be 2, which then fails in airflow
+disconnect_sql = """
+  sudo su - postgres bash -c 'echo "select pg_terminate_backend(pid) from pg_stat_activity where datname=\'manifold\' AND application_name!=\'psql\';" > disconnect_db.sql'
```

Comment:
Two things
1. Multiline for readability please
1. Be more specific about where the `disconnect_db.sql` gets written. As of right now, I could not predict where to find it. Maybe put it in /tmp/disconnect_db.sql

Reply:
Cleaned up comments.

Moved SQL script to /tmp/ with a clearer name.